File: Chapter3/Chapter3/Renderer.m
//
//  Renderer.m
//  Chapter3
//
//  Created by Jinwoo Kim on 8/26/24.
//

#import "Renderer.h"

@interface Renderer () <MTKViewDelegate>
@property (retain, nonatomic, nullable) id<MTLDevice> device;
@property (retain, nonatomic, nullable) id<MTLCommandQueue> commandQueue;
@property (retain, nonatomic, nullable) id<MTLLibrary> library;
@property (retain, nonatomic, nullable) MTKMesh *mesh;
@property (retain, nonatomic, nullable) id<MTLBuffer> vertexBuffer;
@property (retain, nonatomic, nullable) id<MTLRenderPipelineState> piplineState;
@end

@implementation Renderer

- (instancetype)initWithMetalView:(MTKView *)metalView {
    if (self = [super init]) {
        metalView.delegate = self;
        metalView.clearColor = MTLClearColorMake(1., 1., 0.8, 1.);
        metalView.depthStencilPixelFormat = MTLPixelFormatInvalid;
        
        //
        
        id<MTLDevice> device = MTLCreateSystemDefaultDevice();
        id<MTLCommandQueue> commandQueue = [device newCommandQueue];
        
        metalView.device = device;
        
        MTKMeshBufferAllocator *allocator = [[MTKMeshBufferAllocator alloc] initWithDevice:device];
        MDLMesh *mdlMesh = [[MDLMesh alloc] initBoxWithExtent:simd_make_float3(0.8, 0.8, 0.8)
                                                     segments:simd_make_uint3(1, 1, 1)
                                                inwardNormals:NO
                                                 geometryType:MDLGeometryTypeTriangles
                                                    allocator:allocator];
        [allocator release];
        
        NSError * _Nullable error = nil;
        MTKMesh *mesh = [[MTKMesh alloc] initWithMesh:mdlMesh device:device error:&error];
        assert(error == nil);
        
        id<MTLLibrary> library = [device newDefaultLibrary];
        id<MTLFunction> vertexFunction = [library newFunctionWithName:@"vertex_main"];
        id<MTLFunction> fragmentFunction = [library newFunctionWithName:@"fragment_main"];
        
        MTLRenderPipelineDescriptor *piplineDescriptor = [MTLRenderPipelineDescriptor new];
        piplineDescriptor.vertexFunction = vertexFunction;
        [vertexFunction release];
        piplineDescriptor.fragmentFunction = fragmentFunction;
        [fragmentFunction release];
        piplineDescriptor.colorAttachments[0].pixelFormat = metalView.colorPixelFormat;
        piplineDescriptor.vertexDescriptor = MTKMetalVertexDescriptorFromModelIO(mdlMesh.vertexDescriptor);
        [mdlMesh release];
        
        id<MTLRenderPipelineState> pipelineState = [device newRenderPipelineStateWithDescriptor:piplineDescriptor error:&error];
        [piplineDescriptor release];
        assert(error == nil);
        
        id<MTLBuffer> vertexBuffer = mesh.vertexBuffers[0].buffer;
        
        //
        
        self.device = device;
        self.commandQueue = commandQueue;
        self.mesh = mesh;
        self.vertexBuffer = vertexBuffer;
        self.piplineState = pipelineState;
        
        [device release];
        [commandQueue release];
        [mesh release];
        [pipelineState release];
    }
    
    return self;
}

- (void)dealloc {
    [_device release];
    [_commandQueue release];
    [_library release];
    [_mesh release];
    [_vertexBuffer release];
    [_piplineState release];
    [super dealloc];
}

- (void)mtkView:(MTKView *)view drawableSizeWillChange:(CGSize)size {
    
}

- (void)drawInMTKView:(MTKView *)view {
    id<MTLCommandBuffer> commandBuffer = [self.commandQueue commandBuffer];
    MTLRenderPassDescriptor *descriptor = [view currentRenderPassDescriptor];
    id<MTLRenderCommandEncoder> renderEncoder = [commandBuffer renderCommandEncoderWithDescriptor:descriptor];
    
    //
    
    [renderEncoder setRenderPipelineState:self.piplineState];
    [renderEncoder setVertexBuffer:self.vertexBuffer offset:0 atIndex:0];
    
    for (MTKSubmesh *submesh in self.mesh.submeshes) {
        [renderEncoder drawIndexedPrimitives:MTLPrimitiveTypeTriangle
                                  indexCount:submesh.indexCount
                                   indexType:submesh.indexType
                                 indexBuffer:submesh.indexBuffer.buffer
                           indexBufferOffset:submesh.indexBuffer.offset];
    }
    
    //
    
    [renderEncoder endEncoding];
    [commandBuffer presentDrawable:view.currentDrawable];
    [commandBuffer commit];
}

@end

File: Chapter3/Chapter3/Renderer.swift
//
//  Renderer.swift
//  Chapter3
//
//  Created by Jinwoo Kim on 7/17/22.
//

import Foundation
import MetalKit

class Renderer: NSObject {
    static var device: MTLDevice!
    static var commandQueue: MTLCommandQueue!
    var mesh: MTKMesh!
    var vertexBuffer: MTLBuffer!
    var pipelineState: MTLRenderPipelineState!
    var timer: Float = .zero
    
    var uniforms: Uniforms = .init()
    
    init(mtkView: MTKView) {
        guard let device: MTLDevice = MTLCreateSystemDefaultDevice(),
              let commandQueue: MTLCommandQueue = device.makeCommandQueue() else {
            fatalError("GPU not available")
        }
        
        Self.device = device
        Self.commandQueue = commandQueue
        mtkView.device = device
        
        let library: MTLLibrary = device.makeDefaultLibrary()!
        let vertexFunction: MTLFunction = library.makeFunction(name: "vertex_main")!
        let fragmentFunction: MTLFunction = library.makeFunction(name: "fragment_main")!
        
        //
        
        super.init()
        
        //
        
        mtkView.clearColor = MTLClearColor(red: 1.0, green: 1.0, blue: 0.8, alpha: 1.0)
        mtkView.delegate = self
        
        let mdlMesh: MDLMesh = Primitive.makeCube(device: device, size: 1)
        mesh = try! .init(mesh: mdlMesh, device: device)
        
        vertexBuffer = mesh.vertexBuffers[0].buffer
        
        let pipelineDescriptor: MTLRenderPipelineDescriptor = .init()
        pipelineDescriptor.vertexFunction = vertexFunction
        pipelineDescriptor.fragmentFunction = fragmentFunction
        pipelineDescriptor.vertexDescriptor = try! MTKMetalVertexDescriptorFromModelIOWithError(mdlMesh.vertexDescriptor)
        pipelineDescriptor.colorAttachments[0].pixelFormat = mtkView.colorPixelFormat
        pipelineState = try! device.makeRenderPipelineState(descriptor: pipelineDescriptor)
        
        //
        
        let translation: float4x4 = .init(translation: [0, 0.3, 0])
        let rotation: float4x4 = .init(rotation: [0, Float(45).degreesToRadians, 0])
        uniforms.modelMatrix = translation * rotation
        uniforms.viewMatrix = float4x4(translation: [0.8, 0, 0]).inverse
    }
}

extension Renderer: MTKViewDelegate {
    func mtkView(_ view: MTKView, drawableSizeWillChange size: CGSize) {
        let aspect: Float = Float(view.bounds.width) / Float(view.bounds.height)
        let projectionMatrix: float4x4 = .init(projectionFov: Float(45).degreesToRadians, near: 0.1, far: 100, aspect: aspect)
        uniforms.projectionMatrix = projectionMatrix
    }
    
    func draw(in view: MTKView) {
        print(#function, Date())
        
        let descriptor: MTLRenderPassDescriptor = view.currentRenderPassDescriptor!
        let commandBuffer: MTLCommandBuffer = Self.commandQueue.makeCommandBuffer()!
        let renderEncoder: MTLRenderCommandEncoder = commandBuffer.makeRenderCommandEncoder(descriptor: descriptor)!
        
        //
        
//        timer += 0.05
//        var currentTime: Float = sin(timer)
//        renderEncoder.setVertexBytes(&currentTime, length: MemoryLayout<Float>.stride, index: 1)
        
        //
        
        renderEncoder.setVertexBytes(&uniforms, length: MemoryLayout<Uniforms>.stride, index: 1)
        
        timer += 0.05
        uniforms.viewMatrix = float4x4.identity()
        uniforms.modelMatrix = float4x4(rotationY: sin(timer))
        
        uniforms.viewMatrix = float4x4(translation: [0, 0, -3]).inverse
        
        //
        
        renderEncoder.setRenderPipelineState(pipelineState)
        renderEncoder.setVertexBuffer(vertexBuffer, offset: 0, index: 0)
        mesh.submeshes.forEach { submesh in
            renderEncoder.drawIndexedPrimitives(type: .triangle,
                                                indexCount: submesh.indexCount,
                                                indexType: submesh.indexType,
                                                indexBuffer: submesh.indexBuffer.buffer,
                                                indexBufferOffset: submesh.indexBuffer.offset)
        }
        
        renderEncoder.endEncoding()
        let drawable: MTLDrawable = view.currentDrawable!
        commandBuffer.present(drawable)
        commandBuffer.commit()
    }
}
